File: src/cli/query/count.ts
import { Command, ExactArgs, ref } from "vipvot";
import { printJson, printError } from "../../lib/output.ts";
import { getMongoClient, closeAllClients } from "../../mongo/client.ts";
import { countDocuments } from "../../mongo/query.ts";
import { enhanceErrorMessage } from "../../lib/errors.ts";
import { parseJson } from "../../lib/parse-json.ts";
import { resolveConnectionAlias } from "../_globals.ts";

export function buildCountCommand(): Command {
  const filter = ref<string>("");

  const cmd = Command({
    use: "count <database> <collection>",
    short: "Count documents matching a filter",
    args: ExactArgs(2),
    run: async (_cmd, args) => {
      const [database, collection] = args as [string, string];
      try {
        const { client } = await getMongoClient(resolveConnectionAlias());
        const filterDoc = filter.value ? parseJson(filter.value, "filter") : undefined;
        const count = await countDocuments(client, {
          dbName: database,
          collName: collection,
          filter: filterDoc,
        });
        printJson({ database, collection, filter: filterDoc ?? {}, count });
      } catch (err) {
        printError(
          err instanceof Error
            ? enhanceErrorMessage(err, { database, collection })
            : "Failed to count documents",
        );
      } finally {
        await closeAllClients();
      }
    },
  });

  cmd.flags().stringVarP(filter, "filter", "", "", "MongoDB query filter (JSON)");

  return cmd;
}

File: internal/cli/query/count.go
package query

import (
	"github.com/spf13/cobra"
	"go.mongodb.org/mongo-driver/v2/bson"

	"github.com/shhac/agent-mongo/internal/cli/shared"
	"github.com/shhac/agent-mongo/internal/mongo"
	"github.com/shhac/agent-mongo/internal/output"
)

func registerCount(parent *cobra.Command, globals func() *shared.GlobalFlags) {
	var filter string

	cmd := &cobra.Command{
		Use:   "count <database> <collection>",
		Short: "Count documents matching a filter",
		Args:  cobra.ExactArgs(2),
		RunE: func(_ *cobra.Command, args []string) error {
			g := globals()
			ref := mongo.Ref{DB: args[0], Collection: args[1]}
			filterDoc, err := parseOptionalDoc(filter, "filter")
			if err != nil {
				return err
			}
			return shared.WithSessionRef(g, ref, func(ctx shared.SessionCtx) error {
				count, err := ctx.Session.CountDocuments(ctx.Ctx, ref, filterDoc)
				if err != nil {
					return err
				}
				return printCount(count, ref, filterDoc)
			})
		},
	}

	cmd.Flags().StringVar(&filter, "filter", "", "MongoDB query filter (JSON)")
	parent.AddCommand(cmd)
}

func printCount(count int64, ref mongo.Ref, filter bson.D) error {
	var e echo
	e.doc("filter", filter)
	return output.PrintResultWithMeta(map[string]any{
		"database":   ref.DB,
		"collection": ref.Collection,
		"count":      count,
	}, e.meta())
}
